File: examples/C/flclient1.c
//
//  Freelance client - Model 1
//  Uses REQ socket to query one or more services
//
#include "czmq.h"

#define REQUEST_TIMEOUT     1000
#define MAX_RETRIES         3       //  Before we abandon

static zmsg_t *
s_try_request (zctx_t *ctx, char *endpoint, zmsg_t *request)
{
    printf ("I: trying echo service at %s...\n", endpoint);
    void *client = zsocket_new (ctx, ZMQ_REQ);
    zsocket_connect (client, endpoint);

    //  Send request, wait safely for reply
    zmsg_t *msg = zmsg_dup (request);
    zmsg_send (&msg, client);
    zmq_pollitem_t items [] = { { client, 0, ZMQ_POLLIN, 0 } };
    zmq_poll (items, 1, REQUEST_TIMEOUT * ZMQ_POLL_MSEC);
    zmsg_t *reply = NULL;
    if (items [0].revents & ZMQ_POLLIN)
        reply = zmsg_recv (client);

    //  Close socket in any case, we're done with it now
    zsocket_destroy (ctx, client);
    return reply;
}

//  .split client task
//  The client uses a Lazy Pirate strategy if it only has one server to talk
//  to. If it has 2 or more servers to talk to, it will try each server just
//  once:

int main (int argc, char *argv [])
{
    zctx_t *ctx = zctx_new ();
    zmsg_t *request = zmsg_new ();
    zmsg_addstr (request, "Hello world");
    zmsg_t *reply = NULL;

    int endpoints = argc - 1;
    if (endpoints == 0)
        printf ("I: syntax: %s <endpoint> ...\n", argv [0]);
    else
    if (endpoints == 1) {
        //  For one endpoint, we retry N times
        int retries;
        for (retries = 0; retries < MAX_RETRIES; retries++) {
            char *endpoint = argv [1];
            reply = s_try_request (ctx, endpoint, request);
            if (reply)
                break;          //  Successful
            printf ("W: no response from %s, retrying...\n", endpoint);
        }
    }
    else {
        //  For multiple endpoints, try each at most once
        int endpoint_nbr;
        for (endpoint_nbr = 0; endpoint_nbr < endpoints; endpoint_nbr++) {
            char *endpoint = argv [endpoint_nbr + 1];
            reply = s_try_request (ctx, endpoint, request);
            if (reply)
                break;          //  Successful
            printf ("W: no response from %s\n", endpoint);
        }
    }
    if (reply)
        printf ("Service is running OK\n");

    zmsg_destroy (&request);
    zmsg_destroy (&reply);
    zctx_destroy (&ctx);
    return 0;
}

File: examples/Java/flclient1.java
package guide;

import org.zeromq.SocketType;
import org.zeromq.ZContext;
import org.zeromq.ZMQ;
import org.zeromq.ZMQ.Poller;
import org.zeromq.ZMQ.Socket;
import org.zeromq.ZMsg;

//  Freelance client - Model 1
//  Uses REQ socket to query one or more services
public class flclient1
{
    private static final int REQUEST_TIMEOUT = 1000;
    private static final int MAX_RETRIES     = 3;   //  Before we abandon

    private static ZMsg tryRequest(ZContext ctx, String endpoint, ZMsg request)
    {
        System.out.printf("I: trying echo service at %s...\n", endpoint);
        Socket client = ctx.createSocket(SocketType.REQ);
        client.connect(endpoint);

        //  Send request, wait safely for reply
        ZMsg msg = request.duplicate();
        msg.send(client);

        Poller poller = ctx.createPoller(1);
        poller.register(client, Poller.POLLIN);

        poller.poll(REQUEST_TIMEOUT);

        ZMsg reply = null;
        if (poller.pollin(0))
            reply = ZMsg.recvMsg(client);

        //  Close socket in any case, we're done with it now
        ctx.destroySocket(client);
        poller.close();
        return reply;
    }
    //  .split client task
    //  The client uses a Lazy Pirate strategy if it only has one server to talk
    //  to. If it has two or more servers to talk to, it will try each server just
    //  once:

    public static void main(String[] argv)
    {
        try (ZContext ctx = new ZContext()) {
            ZMsg request = new ZMsg();
            request.add("Hello world");
            ZMsg reply = null;

            int endpoints = argv.length;
            if (endpoints == 0)
                System.out.printf("I: syntax: flclient1 <endpoint> ...\n");
            else if (endpoints == 1) {
                //  For one endpoint, we retry N times
                int retries;
                for (retries = 0; retries < MAX_RETRIES; retries++) {
                    String endpoint = argv[0];
                    reply = tryRequest(ctx, endpoint, request);
                    if (reply != null)
                        break; //  Successful
                    System.out.printf(
                        "W: no response from %s, retrying...\n", endpoint
                    );
                }
            }
            else {
                //  For multiple endpoints, try each at most once
                int endpointNbr;
                for (endpointNbr = 0; endpointNbr < endpoints; endpointNbr++) {
                    String endpoint = argv[endpointNbr];
                    reply = tryRequest(ctx, endpoint, request);
                    if (reply != null)
                        break; //  Successful
                    System.out.printf("W: no response from %s\n", endpoint);
                }
            }
            if (reply != null) {
                System.out.printf("Service is running OK\n");
                reply.destroy();
            }
            request.destroy();
        }
    }

}
